framework/set/resources/providers/harvester/instances.go

```diff
@@ -0,0 +1,167 @@
+package harvester
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/rancher/shepherd/pkg/namegenerator"
+	"github.com/rancher/tfp-automation/config"
+	"github.com/rancher/tfp-automation/framework/set/defaults"
+	"github.com/zclconf/go-cty/cty"
+	"golang.org/x/crypto/ssh"
+)
+
+// getPublicSSHKey gets the public key from a private SSH key and returns it as a string
+func getPublicSSHKey(privateKeyPath string) string {
+	sshBytes, err := os.ReadFile(privateKeyPath)
+	if err != nil {
+		log.Fatalf("Failed to read private key file: %v", err)
+	}
+
+	signer, err := ssh.ParsePrivateKey(sshBytes)
+	if err != nil {
+		log.Fatalf("Failed to parse OpenSSH private key: %v", err)
+		return ""
+	}
+
+	publicKey := signer.PublicKey()
+
+	// Convert the public key to authorized_keys format
+	pubKeyString := string(ssh.MarshalAuthorizedKey(publicKey))
+
+	return pubKeyString
+}
+
+// CreateHarvesterInstances is a function that will set the Harvester instances configurations in the main.tf file.
+func CreateHarvesterInstances(rootBody *hclwrite.Body, terraformConfig *config.TerraformConfig, terratestConfig *config.TerratestConfig,
+	hostnamePrefix string) {
+
+	configBlockSSHKey := rootBody.AppendNewBlock(defaults.Resource, []string{defaults.HarvesterSSHKey, hostnamePrefix + "ssh_key"})
+	configBlockSSHKeyBody := configBlockSSHKey.Body()
+
+	configBlockSSHKeyBody.SetAttributeValue(defaults.Name, cty.StringVal(namegenerator.AppendRandomString("tfpsshkey")))
+	configBlockSSHKeyBody.SetAttributeValue(defaults.Namespace, cty.StringVal(terraformConfig.HarvesterConfig.VMNamespace))
+
+	publicKey := getPublicSSHKey(terraformConfig.PrivateKeyPath)
+	configBlockSSHKeyBody.SetAttributeValue(defaults.PublicKey, cty.StringVal(publicKey))
+
+	configBlockSecret := rootBody.AppendNewBlock(defaults.Resource, []string{defaults.KubernetesSecret, hostnamePrefix + "secret"})
+	configBlockSecretBody := configBlockSecret.Body()
+
+	secretBlockMeta := configBlockSecretBody.AppendNewBlock("metadata", []string{})
+	secretBlockMetaBody := secretBlockMeta.Body()
+
+	secretName := namegenerator.AppendRandomString("tfpsecret")
+
+	secretBlockMetaBody.SetAttributeValue(defaults.Name, cty.StringVal(secretName))
+	secretBlockMetaBody.SetAttributeValue(defaults.Namespace, cty.StringVal(terraformConfig.HarvesterConfig.VMNamespace))
+
+	secretBlockMetaBody.SetAttributeValue(defaults.Labels, cty.ObjectVal(map[string]cty.Value{
+		"sensitive": cty.StringVal("false"),
+	}))
+
+	hclLocalValue := hclwrite.Tokens{
+		{Type: hclsyntax.TokenIdent, Bytes: []byte("{\"userdata\" = local." + defaults.CloudInit + "}")},
+	}
+
+	configBlockSecretBody.SetAttributeRaw("data", hclLocalValue)
+
+	configBlock := rootBody.AppendNewBlock(defaults.Resource, []string{defaults.HarvesterVirtualMachine, hostnamePrefix})
+	configBlockBody := configBlock.Body()
+
+	// not sure what this does yet, but it was for aws
```

Comment:
This specifically when doing tests in Rancher that utilize custom clusters. This is in for Linode as well, should be fine to keep here, just remove the comment.